File: packages/gg-sandbox-swift/Sources/SDK/Modules/Files.swift
/// Read, write, edit, list, walk and search the files of the workspace.
///
/// Nothing here places material in the context window.
///
/// - ggmodule: files
public enum files {
    /// The gg tools this module dispatches, which is its share of what the artifact answers
    /// `bound-operations` with. Declared beside the functions that call them, so a tool added here
    /// is a tool the artifact reports.
    static let ggOperations = [
        "read_file", "write_file", "edit_file", "list_dir", "tree", "search",
    ]

    /// Read a file, as either a `FileRead.text` or a `FileRead.image`.
    ///
    /// Which of the two comes back is detected from the file's bytes, never from the extension. The
    /// enum is closed, so a `switch` over it needs no `default`:
    ///
    /// ```swift
    /// switch try files.readFile("logo.png") {
    /// case .text(let file): print(file.contents)
    /// case .image(let picture): print(picture.label)
    /// }
    /// ```
    ///
    /// A relative path resolves against the workspace; an absolute one is read as given, so
    /// anything else in this container — an offloaded command's output under `/tmp/gg-shell`, say —
    /// is readable. The bytes go to the program and nothing is placed in the context window; a
    /// picture is described rather than shown.
    ///
    /// - Parameters:
    ///   - path: The file to read, relative to the workspace or absolute.
    ///   - offset: The 1-based line to start at. Left out, the read starts at the first line.
    ///   - limit: How many lines to return from `offset`. Left out, the read runs to the end.
    /// - Returns: the file's text window, or the picture's description.
    /// - Throws: `core.ApiError` with `.notFound` for a missing path.
    /// - ggop: files.read_file
    public static func readFile(
        _ path: String, offset: Int? = nil, limit: Int? = nil
    ) throws -> FileRead {
        try withScratch { scratch in
            var path = scratch.string(path)
            var ret = test_cabinet_gg_files_file_read_t()
            var err = test_cabinet_gg_types_api_error_t()
            let ok = withWindow(offset, limit) { offset, limit in
                test_cabinet_gg_files_read_file(&path, offset, limit, &ret, &err)
            }
            guard ok else { throw lift(failure: &err) }
            let read = FileRead(wire: ret)
            test_cabinet_gg_files_file_read_free(&ret)
            return read
        }
    }

    /// Write UTF-8 text to a file, creating parent directories and replacing what is there.
    ///
    /// - Parameters:
    ///   - path: Where to write, relative to the workspace or absolute. Parent directories are
    ///     created.
    ///   - contents: The UTF-8 text to write. It replaces the file entirely.
    /// - Returns: how many bytes were written.
    /// - Throws: `core.ApiError` with `.invalidArgument` for an empty path, and `.ioError` when
    ///   creating the parent directories or the write itself failed.
    /// - ggop: files.write_file
    @discardableResult
    public static func writeFile(_ path: String, contents: String) throws -> Int {
        try withScratch { scratch in
            var path = scratch.string(path)
            var contents = scratch.string(contents)
            var ret: UInt64 = 0
            var err = test_cabinet_gg_types_api_error_t()
            guard test_cabinet_gg_files_write_file(&path, &contents, &ret, &err) else {
                throw lift(failure: &err)
            }
            return Int(ret)
        }
    }

    /// Replace the one exact occurrence of some text in a file with something else.
    ///
    /// - Parameters:
    ///   - path: The file to edit.
    ///   - replacing: The exact text to find, whitespace included. It must appear exactly once.
    ///   - with: The text to put in its place. An empty string deletes the match.
    /// - Throws: `core.ApiError` with `.notFound` when the text does not appear, and `.conflict` —
    ///   with the number of matches — when it appears more than once.
    /// - ggop: files.edit_file
    public static func editFile(
        _ path: String, replacing oldString: String, with newString: String
    ) throws {
        try withScratch { scratch in
            var path = scratch.string(path)
            var oldString = scratch.string(oldString)
            var newString = scratch.string(newString)
            var err = test_cabinet_gg_types_api_error_t()
            guard test_cabinet_gg_files_edit_file(&path, &oldString, &newString, &err) else {
                throw lift(failure: &err)
            }
        }
    }

    /// List a directory, sorted by name; leaving the path out lists the workspace root.
    ///
    /// Each entry carries a bare `name` — join it with the directory that was listed — and its
    /// `kind`. An empty directory is an empty array, not a failure.
    ///
    /// - Parameter path: The directory to list, relative to the workspace or absolute. Left out, it
    ///   lists the workspace root.
    /// - Returns: the directory's entries, sorted by name.
    /// - Throws: `core.ApiError` with `.notFound` for a directory that is not there, and
    ///   `.invalidArgument` for a path that is given but empty.
    /// - ggop: files.list_dir
    public static func listDir(_ path: String? = nil) throws -> [DirEntry] {
        try withScratch { scratch in
            var ret = test_cabinet_gg_files_list_dir_entry_t()
            var err = test_cabinet_gg_types_api_error_t()
            let ok = withOptional(path.map { scratch.string($0) }) { path in
                test_cabinet_gg_files_list_dir(path, &ret, &err)
            }
            guard ok else { throw lift(failure: &err) }
            let entries = lift(ret.ptr, ret.len) { DirEntry(wire: $0) }
            test_cabinet_gg_files_list_dir_entry_free(&ret)
            return entries
        }
    }

    /// Render the tree beneath a directory, skipping everything the ignore files exclude.
    ///
    /// One block of text: the root itself unnamed, each level indented two further spaces than its
    /// parent, every level in path order, and directories suffixed `/`. A root with nothing beneath
    /// it renders as `(empty directory)`.
    ///
    /// `depth` counts levels of children below the root, so `1` is the root's own entries. A
    /// directory sitting at the bound is suffixed with how many entries it holds that were not
    /// walked, as `assets/ (12 entries not shown)`.
    ///
    /// Whatever `.gitignore`, `.ignore`, `.git/info/exclude` and the global ignore file exclude —
    /// nested files and negations included — is never walked and never rendered, `.git` itself is
    /// skipped, dotfiles are rendered, symbolic links are not followed, and none of it needs a
    /// repository to be there.
    ///
    /// The rendering is bounded at 1000 lines and 16 KiB, whichever binds first, and a result cut
    /// by either ends with a line saying so.
    ///
    /// - Parameters:
    ///   - path: The directory to walk, relative to the workspace or absolute. Left out, the
    ///     workspace root is walked.
    ///   - depth: How many levels of children below the root to render: 2 when left out, 10 at most
    ///     — a larger request is answered at 10 — and anything below `1` is refused.
    /// - Returns: the rendered tree.
    /// - Throws: `core.ApiError` with `.notFound` for a path that is not there, and
    ///   `.invalidArgument` for a path that is not a directory or a depth below `1`.
    /// - ggop: files.tree
    public static func tree(path: String? = nil, depth: Int? = nil) throws -> String {
        if let depth, depth < 1 {
            throw core.ApiError(
                code: .invalidArgument, operation: "tree",
                message: "depth must be at least 1 (\(depth) given); leave it out for gg's default")
        }
        return try withScratch { scratch in
            var ret = sandbox_string_t()
            var err = test_cabinet_gg_types_api_error_t()
            let ok = withOptional(path.map { scratch.string($0) }) { path in
                withOptional(depth.map { UInt32(clamping: $0) }) { depth in
                    test_cabinet_gg_files_tree(path, depth, &ret, &err)
                }
            }
            guard ok else { throw lift(failure: &err) }
            let rendered = lift(ret)
            sandbox_string_free(&ret)
            return rendered
        }
    }

    /// Search the workspace's files for a regular expression, and hand back every line that matches.
    ///
    /// This is the sandbox's grep. `query` is a regular expression in Rust's syntax — `foo|bar`,
    /// `fn [a-z_]+`, `(?i)todo` for a case-insensitive match — matched against each line on its
    /// own. Every line it matches comes back as a `SearchMatch` carrying the file's path, the
    /// 1-based line number and the line itself, in path order and then line order.
    ///
    /// Whatever `.gitignore`, `.ignore`, `.git/info/exclude` and the global ignore file exclude —
    /// nested files and negations included — is never scanned, `.git` itself is skipped, dotfiles
    /// are searched, and none of it needs a repository to be there. A file carrying a NUL byte is
    /// skipped.
    ///
    /// A matching line longer than 200 characters is cut there and annotated in place as
    /// `foo (123 more chars...)`. There is no offset: an array exactly `limit` long may have been
    /// cut.
    ///
    /// - Parameters:
    ///   - query: The regular expression to match each line against, in Rust's syntax; `(?i)`
    ///     makes it case-insensitive.
    ///   - path: The directory or file to search, relative to the workspace or absolute. Left out,
    ///     the whole workspace is searched; a file searches that one file.
    ///   - limit: How many matches to return at most. Left out, it takes gg's default of 50; the
    ///     ceiling is 200, and a larger limit is clamped to it rather than refused.
    /// - Returns: every matching line up to the limit, in path order and then line order; nothing
    ///   matching is an empty array, not a failure.
    /// - Throws: `core.ApiError` with `.invalidArgument` for a blank query, one that is not a valid
    ///   pattern, or a limit of `0`, and `.notFound` for a path that is not there.
    /// - ggop: files.search
    public static func search(
        _ query: String, path: String? = nil, limit: Int? = nil
    ) throws -> [SearchMatch] {
        try withScratch { scratch in
            var query = scratch.string(query)
            var ret = test_cabinet_gg_files_list_search_match_t()
            var err = test_cabinet_gg_types_api_error_t()
            let ok = withOptional(path.map { scratch.string($0) }) { path in
                withOptional(limit.map { UInt32(truncatingIfNeeded: $0) }) { limit in
                    test_cabinet_gg_files_search(&query, path, limit, &ret, &err)
                }
            }
            guard ok else { throw lift(failure: &err) }
            let matches = lift(ret.ptr, ret.len) { SearchMatch(wire: $0) }
            test_cabinet_gg_files_list_search_match_free(&ret)
            return matches
        }
    }

    /// What a read returned: a text file's window, or a picture's description.
    ///
    /// Image bytes never enter the program; gg attaches the picture to the turn instead.
    public enum FileRead: Sendable {
        /// This file is text.
        case text(TextFile)
        /// This file is a picture, which gg shows rather than handing over its bytes.
        case image(ImageFile)

        init(wire: test_cabinet_gg_files_file_read_t) {
            if Int32(wire.tag) == TEST_CABINET_GG_FILES_FILE_READ_IMAGE {
                self = .image(ImageFile(wire: wire.val.image))
            } else {
                self = .text(TextFile(wire: wire.val.text))
            }
        }
    }

    /// A text file's window, as the `FileRead.text` case carries it.
    public struct TextFile: Sendable {
        /// The file's text, or just the requested window where the read named one.
        public let contents: String
        /// The 1-based first line returned.
        public let firstLine: Int
        /// The 1-based last line returned.
        public let lastLine: Int
        /// The file's total line count.
        public let totalLines: Int
        /// Whether a 256 KiB byte ceiling cut the returned text.
        public let byteTruncated: Bool

        init(wire: test_cabinet_gg_files_text_read_t) {
            contents = lift(wire.contents)
            firstLine = Int(wire.first_line)
            lastLine = Int(wire.last_line)
            totalLines = Int(wire.total_lines)
            byteTruncated = wire.byte_truncated
        }
    }

    /// A picture's description, as the `FileRead.image` case carries it.
    public struct ImageFile: Sendable {
        /// The IANA media type (`image/png`, `image/jpeg`, `image/gif`, `image/webp`).
        public let mediaType: String
        /// The short format label (`PNG`, `JPEG`, `GIF`, `WebP`).
        public let label: String
        /// The file's size in bytes.
        public let bytes: Int
        /// Whether the picture is being attached to this turn to be looked at.
        public let shown: Bool
        /// Why it is not being shown; `nil` when it is.
        public let notShownReason: String?

        init(wire: test_cabinet_gg_files_image_read_t) {
            mediaType = lift(wire.media_type)
            label = lift(wire.label)
            bytes = Int(wire.bytes)
            shown = wire.shown
            notShownReason = lift(wire.not_shown_reason)
        }
    }

    /// One entry `files.listDir` found.
    public struct DirEntry: Sendable {
        /// The entry's bare name, with no directory part. Join it with the directory that was
        /// listed.
        public let name: String
        /// What the entry is.
        public let kind: EntryKind

        init(wire: test_cabinet_gg_files_dir_entry_t) {
            name = lift(wire.name)
            kind = EntryKind(wire: wire.kind)
        }
    }

    /// One line `files.search` matched.
    public struct SearchMatch: Sendable {
        /// The file's path, relative to the workspace root, with `/` separators.
        ///
        /// Absolute for a search rooted outside the workspace.
        public let path: String
        /// The 1-based line number of the match within that file.
        public let line: Int
        /// The matching line, without its line ending.
        ///
        /// Longer than 200 characters, it is cut there and annotated in place as
        /// `foo (123 more chars...)`.
        public let text: String

        init(wire: test_cabinet_gg_files_search_match_t) {
            path = lift(wire.path)
            line = Int(wire.line)
            text = lift(wire.text)
        }
    }

    /// What a directory entry is.
    public enum EntryKind: Sendable {
        /// An ordinary file.
        case file
        /// A directory, which can be listed in turn.
        case directory
        /// Everything that is neither, a symlink among them.
        case other

        init(wire: test_cabinet_gg_files_entry_kind_t) {
            switch Int32(wire) {
            case TEST_CABINET_GG_FILES_ENTRY_KIND_FILE: self = .file
            case TEST_CABINET_GG_FILES_ENTRY_KIND_DIRECTORY: self = .directory
            default: self = .other
            }
        }
    }
}

/// The `offset`/`limit` pair the three reads share, as the two nullable pointers the ABI takes.
///
/// Not model-facing: a program writes `offset:`/`limit:` and never sees this.
@inline(__always)
func withWindow<R>(
    _ offset: Int?, _ limit: Int?,
    _ body: (UnsafeMutablePointer<UInt32>?, UnsafeMutablePointer<UInt32>?) throws -> R
) rethrows -> R {
    try withOptional(offset.map { UInt32(truncatingIfNeeded: $0) }) { offset in
        try withOptional(limit.map { UInt32(truncatingIfNeeded: $0) }) { limit in
            try body(offset, limit)
        }
    }
}

File: packages/gg-sandbox-ruby/src/gg/files.rb
# frozen_string_literal: true

module GG
  # Read, write, edit, list and walk the files of the workspace.
  #
  # Nothing here places anything in the context window.
  module Files
    extend Surface::Operations

    # Read a file, as either a `GG::Files::TextFile` or a `GG::Files::ImageFile`.
    #
    # Which of the two comes back is detected from the file's bytes, never from the extension. Both
    # are classes, so a `case` narrows them.
    #
    # A relative path resolves against the workspace; an absolute one is read as given, so anything
    # else in this container — an offloaded command's output under `/tmp/gg-shell`, say — is
    # readable.
    #
    # The bytes go to the program and nothing is placed in the context window. A picture is
    # described rather than shown.
    #
    # @param path [String] The file to read, relative to the workspace or absolute.
    # @param offset [Integer, nil] The 1-based line to start at. Left out, the read starts at the
    #   first line.
    # @param limit [Integer, nil] How many lines to return from `offset`. Left out, a capped read
    #   policy's default applies, or the read runs to the end of the file.
    # @return [GG::Files::TextFile, GG::Files::ImageFile] the file's window of text, or the
    #   picture's description
    # @raise [GG::Core::ApiError] `:not_found` for a missing path.
    def self.read_file(path, offset: nil, limit: nil)
      Wire.file_read(Wire.call("read_file", "files", "readFile", [
                                 path,
                                 Wire.js(Check.uint("read_file", "offset", offset)),
                                 Wire.js(Check.uint("read_file", "limit", limit))
                               ]))
    end
    operation :read_file, "files.read_file", tool: "read_file"

    # Write UTF-8 text to a file, creating parent directories and replacing what is there.
    #
    # The contents may be given as a `contents` argument or as a block returning the text.
    #
    # @overload write_file(path, contents)
    #   @param path [String] Where to write, relative to the workspace or absolute. Parent
    #     directories are created.
    #   @param contents [String] The UTF-8 text to write. It replaces the file entirely.
    # @overload write_file(path, &contents)
    #   @param path [String] Where to write, relative to the workspace or absolute. Parent
    #     directories are created.
    #   @param contents [String] A block returning the UTF-8 text to write. It replaces the file
    #     entirely.
    # @return [Integer] how many bytes were written
    # @raise [GG::Core::ApiError] `:invalid_argument` when neither a `contents` argument nor a
    #   block was given, and for an empty path; `:io_error` when creating the parent directories or
    #   the write itself failed.
    def self.write_file(path, contents = nil, &block)
      text = block ? block.call : contents
      if text.nil?
        raise Core::ApiError.new("write_file", Core::ApiErrorCode::INVALID_ARGUMENT,
                                 "`write_file` needs the text to write, as an argument or as a " \
                                 "block")
      end

      Wire.integer(Wire.call("write_file", "files", "writeFile", [path, text]))
    end
    operation :write_file, "files.write_file", tool: "write_file"

    # Replace the one exact occurrence of some text in a file with something else.
    #
    # @param path [String] The file to edit.
    # @param old_string [String] The exact text to find, whitespace included. It must appear exactly
    #   once.
    # @param new_string [String] The text to put in its place. An empty string deletes the match.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:not_found` when the text does not appear, and `:conflict` —
    #   with the number of matches — when it appears more than once.
    def self.edit_file(path, old_string, new_string)
      Wire.call("edit_file", "files", "editFile", [path, old_string, new_string])
      nil
    end
    operation :edit_file, "files.edit_file", tool: "edit_file"

    # List a directory, sorted by name; no argument lists the workspace root.
    #
    # Each entry carries a bare `name`, with no directory part, and its `kind`. An empty directory
    # is an empty array, not a failure.
    #
    # @param path [String, nil] The directory to list, relative to the workspace or absolute. Leave
    #   it out for the workspace root.
    # @return [Array<GG::Files::DirEntry>] what the directory holds, sorted by name
    # @raise [GG::Core::ApiError] `:not_found` for a missing directory, and `:invalid_argument` for
    #   a path that is given but empty — leaving it out is what lists the workspace root.
    def self.list_dir(path = nil)
      Wire.call("list_dir", "files", "listDir", [Wire.js(path)]).map do |entry|
        DirEntry.new(
          name: Wire.field(entry, "name"),
          kind: Wire.symbol(`#{entry}.kind`)
        )
      end
    end
    operation :list_dir, "files.list_dir", tool: "list_dir"

    # Render the tree beneath a directory, skipping everything the ignore files exclude.
    #
    # One block of text: the root itself unnamed, each level indented two further spaces than its
    # parent, every level in path order, and directories suffixed `/`. A root with nothing beneath
    # it renders as `(empty directory)`.
    #
    # `depth` counts levels of children below the root, so `1` is the root's own entries. A
    # directory sitting at the bound is suffixed with how many entries it holds that were not
    # walked, as `assets/ (12 entries not shown)`.
    #
    # What `.gitignore`, `.ignore` and their kin exclude — nested files, negations and
    # `.git/info/exclude` included, and `.git` itself — is never walked and never rendered, whether
    # or not the workspace is a repository yet. Dotfiles are otherwise rendered like any other
    # entry, and symbolic links are not followed.
    #
    # The rendering is bounded at 1000 lines and 16 KiB, whichever binds first, and a result cut by
    # either ends with a line saying so.
    #
    # @param path [String, nil] The directory to walk, relative to the workspace or absolute. Leave
    #   it out for the workspace root.
    # @param depth [Integer, nil] How many levels of children below the root to render, at least 1.
    #   Leave it out for 2; the ceiling is 10, and a larger request is answered at 10.
    # @return [String] the rendered tree
    # @raise [GG::Core::ApiError] `:not_found` for a `path` that does not exist, and
    #   `:invalid_argument` for a `path` that is not a directory or a `depth` of zero.
    def self.tree(path: nil, depth: nil)
      bound = Check.uint("tree", "depth", depth)
      if bound == 0
        raise Core::ApiError.new("tree", Core::ApiErrorCode::INVALID_ARGUMENT,
                                 "`depth` must be at least 1, got 0; leave it out for gg's " \
                                 "default of 2")
      end

      Wire.call("tree", "files", "tree", [Wire.js(path), Wire.js(bound)])
    end
    operation :tree, "files.tree", tool: "tree"

    # Search the workspace's files for a regular expression and hand back every matching line.
    #
    # A `grep` over the workspace. `query` is a regular expression — Rust syntax, so `foo|bar`,
    # `fn\s+update`, and `(?i)todo` for a case-insensitive match — tried against each line on its
    # own, and every line it matches comes back with its path and 1-based line number, in path
    # order and then line order. It is the pattern's text, as a String, rather than a Ruby Regexp.
    # `path` roots the search at one directory or one file; leave it out for the workspace root.
    #
    # What `.gitignore`, `.ignore` and their kin exclude — nested files, negations and
    # `.git/info/exclude` included, and `.git` itself — is never scanned and never returned, whether
    # or not the workspace is a repository yet. A file carrying a NUL byte is skipped. Dotfiles are
    # otherwise searched like any other file. A file under an ignored path is still readable by its
    # path.
    #
    # At most `limit` matches come back — 50 by default, and never more than 200 — and a list
    # exactly `limit` long may have been cut. There is no offset. A matching line longer than 200
    # characters is cut there and annotated in place as `foo (123 more chars...)`.
    #
    # @param query [String] The regular expression to match each line against, in Rust syntax;
    #   `(?i)` at the front makes it case-insensitive. It may not be blank.
    # @param path [String, nil] The directory or file to search, relative to the workspace or
    #   absolute. Leave it out to search the whole workspace.
    # @param limit [Integer, nil] The most matches to return, at least 1. Leave it out for 50; the
    #   ceiling is 200, and a larger request is answered with the first 200 rather than refused.
    # @return [Array<GG::Files::SearchMatch>] every matching line, in path order and then line
    #   order, each with its `path`, 1-based `line` and `text`; an empty array when nothing matched,
    #   and one exactly `limit` long may have been cut
    # @raise [GG::Core::ApiError] `:invalid_argument` for a blank query, one that is not a valid
    #   pattern, a `query` that is not a String, or a `limit` of zero, and `:not_found` for a `path`
    #   that does not exist.
    def self.search(query, path: nil, limit: nil)
      unless query.is_a?(String)
        raise Core::ApiError.new("search", Core::ApiErrorCode::INVALID_ARGUMENT,
                                 "`query` is the pattern's text as a String, in Rust regular-" \
                                 "expression syntax, got #{query.inspect}")
      end

      bound = Check.uint("search", "limit", limit)
      if bound == 0
        raise Core::ApiError.new("search", Core::ApiErrorCode::INVALID_ARGUMENT,
                                 "`limit` must be at least 1, got 0; leave it out for gg's " \
                                 "default of 50")
      end

      Wire.call("search", "files", "search", [query, Wire.js(path), Wire.js(bound)]).map do |match|
        SearchMatch.new(
          path: Wire.field(match, "path"),
          line: Wire.field(match, "line"),
          text: Wire.field(match, "text")
        )
      end
    end
    operation :search, "files.search", tool: "search"

    # A text file's window, as `GG::Files.read_file` returns it.
    class TextFile
      include Value

      # @return [String] The file's text, or just the requested window where the read named one.
      attr_reader :contents

      # @return [Integer] The 1-based first line returned.
      attr_reader :first_line

      # @return [Integer] The 1-based last line returned.
      attr_reader :last_line

      # @return [Integer] The file's total line count.
      attr_reader :total_lines

      # @api private
      def initialize(contents:, first_line:, last_line:, total_lines:, byte_truncated:)
        @contents = contents
        @first_line = first_line
        @last_line = last_line
        @total_lines = total_lines
        @byte_truncated = byte_truncated
        freeze
      end

      # @return [Boolean] Whether a 256 KiB byte ceiling cut the returned text.
      def byte_truncated?
        @byte_truncated
      end
    end

    # A picture's description, as `GG::Files.read_file` returns it.
    #
    # The pixels never enter the program.
    class ImageFile
      include Value

      # @return [String] The IANA media type (`image/png`, `image/jpeg`, `image/gif`, `image/webp`).
      attr_reader :media_type

      # @return [String] The short format label (`PNG`, `JPEG`, `GIF`, `WebP`).
      attr_reader :label

      # @return [Integer] The file's size in bytes.
      attr_reader :bytes

      # @return [String, nil] Why the picture is not being shown; `nil` when it is.
      attr_reader :not_shown_reason

      # @api private
      def initialize(media_type:, label:, bytes:, shown:, not_shown_reason:)
        @media_type = media_type
        @label = label
        @bytes = bytes
        @shown = shown
        @not_shown_reason = not_shown_reason
        freeze
      end

      # @return [Boolean] Whether the picture is being attached to this turn to be looked at.
      def shown?
        @shown
      end
    end

    # What a directory entry is.
    #
    # Every arm is a Symbol: `GG::Files::EntryKind::FILE` and `:file` are the same value.
    module EntryKind
      # An ordinary file.
      FILE = :file

      # A directory, which can be listed in its own right.
      DIRECTORY = :directory

      # Everything that is neither, a symlink among them.
      OTHER = :other
    end

    # One entry `GG::Files.list_dir` found: a bare name, and its kind.
    class DirEntry
      include Value

      # @return [String] The entry's bare name, with no directory part.
      attr_reader :name

      # @return [GG::Files::EntryKind] What the entry is.
      attr_reader :kind

      # @api private
      def initialize(name:, kind:)
        @name = name
        @kind = kind
        freeze
      end
    end

    # One line `GG::Files.search` matched: where it is, and the line itself.
    class SearchMatch
      include Value

      # @return [String] The file's path, relative to the workspace root — or absolute, for a search
      #   rooted outside it — with `/` separators.
      attr_reader :path

      # @return [Integer] The 1-based line number of the match within that file.
      attr_reader :line

      # The matching line, without its line ending.
      #
      # A line longer than 200 characters is cut there and annotated in place as
      # `foo (123 more chars...)`.
      #
      # @return [String] the matching line
      attr_reader :text

      # @api private
      def initialize(path:, line:, text:)
        @path = path
        @line = line
        @text = text
        freeze
      end
    end
  end
end
